### PR TITLE
Modularization of metrics

### PR DESCRIPTION
**Pull request by dylandreimerink**
This PR restructures the way we work with metrics within Cilium. The goal of this PR is to address a few pain points with the current metrics system. We get three big things out of this work:
* The ability to define metrics in a modular way, close to the location where it is actually going to be used.
* The ability to define richer metadata like label descriptions, known values and their meaning etc.
* The ability to automatically generate metrics references from all defined metrics

To accomplish we did the following things:
* Created custom metric types on top of the existing Prometheus metrics, which are API compatible, but carry additional information and setting used for other features.
* Metrics are now always registered with the repository, but they can be enabled or disabled. Instead of having a centralized switch statement to determine which metrics will be enabled or disabled, the metric definitions now have a field to say if they are enabled by default or not. The main advantage here is that we are able to list all metrics even those who are disabled.
* A new cell was introduced to wire metric definitions into Hive. Metrics are always contained within a struct so this type can be used to inject the references into other cells that need to use them. The cell also makes the individual metrics available so the registry can collect all metrics defined everywhere.
* All metric definitions in `pkg/metrics/metrics.go` are not converted to a single large cell called `LegacyMetrics` and (for now) are written to the same global variables as were used before.
* The registry has been modularized and uses Hive to collect all metrics
* A tool was introduced in `tools/metricdocgen` which can collect all metrics and render the `Documentation/observability/metrics.rst` file with agent metric references generated from the metrics.

Future PRs will focus on removing app-wide dependency on global variables and start moving metric definitions next to the code that uses it, this will likely go in steps as we modularize the rest of Cillium step by step.

### CODE REVIEW

File: Documentation/cmdref/cilium-agent.md [hunk 1]
```diff
@@ -234,7 +234,7 @@ cilium-agent [flags]
       --mesh-auth-mtls-listener-port int                        Port on which the Cilium Agent will perfom mTLS handshakes between other Agents
       --mesh-auth-spiffe-trust-domain string                    The trust domain for the SPIFFE identity. (default "spiffe.cilium.io")
       --mesh-auth-spire-admin-socket string                     The path for the SPIRE admin agent Unix socket.
-      --metrics strings                                         Metrics that should be enabled or disabled from the default metric list. The list is expected to be separated by a space. (+metric_foo to enable metric_foo , -metric_bar to disable metric_bar)
+      --metrics strings                                         Metrics that should be enabled or disabled from the default metric list. (+metric_foo to enable metric_foo , -metric_bar to disable metric_bar)
```
Comment (zacharysarah) [inline Documentation/cmdref/cilium-agent.md:237]: Avoid soft conditional text.
```suggestion
      --metrics strings                                         Metrics to enable or disable from the default metric list. (+metric_foo to enable metric_foo , -metric_bar to disable metric_bar)
```

File: Documentation/cmdref/cilium-agent_hive.md [hunk 2]
```diff
@@ -29,9 +29,11 @@ cilium-agent hive [flags]
       --mesh-auth-mtls-listener-port int       Port on which the Cilium Agent will perfom mTLS handshakes between other Agents
       --mesh-auth-spiffe-trust-domain string   The trust domain for the SPIFFE identity. (default "spiffe.cilium.io")
       --mesh-auth-spire-admin-socket string    The path for the SPIRE admin agent Unix socket.
+      --metrics strings                        Metrics that should be enabled or disabled from the default metric list. (+metric_foo to enable metric_foo , -metric_bar to disable metric_bar)
```
Comment (zacharysarah) [inline Documentation/cmdref/cilium-agent_hive.md:32]: Avoid soft conditional text.

```suggestion
      --metrics strings                        Metrics to enable or disable from the default metric list. (+metric_foo to enable metric_foo , -metric_bar to disable metric_bar)
```

File: Documentation/observability/metrics.rst [hunk 3]
```diff
@@ -219,255 +219,474 @@ disable the following two metrics as they generate too much data:
...
-``node_connectivity_latency_seconds``      ``address_type``, ``protocol``, ``source_cluster``, ``source_node_name``, ``target_cluster``, ``target_node_ip``, ``target_node_name``, ``target_node_type``, ``type`` Enabled    The last observed latency between the current Cilium agent and other Cilium nodes in seconds
-========================================== ====================================================================================================================================================================== ========== ===================================================================================================================
+eBPF
+~~~~
+Both ``bpf_maps_virtual_memory_max_bytes`` and ``bpf_progs_virtual_memory_max_bytes`` are currently reporting the system-wide memory usage of eBPF that is directly and not directly managed by Cilium. This might change in the future and only report the eBPF memory usage directly managed by Cilium.
```
Comment (zacharysarah) [inline Documentation/observability/metrics.rst:287]: Avoid statements about the future. 
```suggestion
Both ``bpf_maps_virtual_memory_max_bytes`` and ``bpf_progs_virtual_memory_max_bytes`` are currently reporting the system-wide memory usage of eBPF that is directly and not directly managed by Cilium.
```

File: Documentation/observability/metrics.rst [hunk 4]
```diff
@@ -219,255 +219,474 @@ disable the following two metrics as they generate too much data:
...
+"""""""""""
+
++-------------+--------------------------------------------------------------------+-----------+-----------------+
+|Name         |Description                                                         |Known Value|Value Description|
++=============+====================================================================+===========+=================+
+|``map_name`` |The label for the BPF map name                                      |           |                 |
++-------------+--------------------------------------------------------------------+-----------+-----------------+
+|``operation``|The label for BPF maps operations                                   |           |                 |
++-------------+--------------------------------------------------------------------+-----------+-----------------+
+|``outcome``  |Indicates whether the outcome of the operation was successful or not|``fail``   |                 |
```
Comment (zacharysarah) [inline Documentation/observability/metrics.rst:309]: Use active voice
```suggestion
|``outcome``  |Indicates whether the operation succeeded or not|``fail``   |                 |
```

File: Documentation/observability/metrics.rst [hunk 5]
```diff
@@ -219,255 +219,474 @@ disable the following two metrics as they generate too much data:
...
++-------------+-----------------------------------------------------------------------------------------------------+-----------+-----------------+
+|``direction``|The label for traffic direction                                                                      |           |                 |
++-------------+-----------------------------------------------------------------------------------------------------+-----------+-----------------+
+|``family``   |Marks which protocol family (IPv4, IPV6) the metric is related to.                                   |           |                 |
++-------------+-----------------------------------------------------------------------------------------------------+-----------+-----------------+
+|``name``     |marks a unique identifier for this metric. The name should be defined once for a given type of error.|           |                 |
++-------------+-----------------------------------------------------------------------------------------------------+-----------+-----------------+
+|``protocol`` |Marks the L4 protocol (TCP, ANY) for the metric.                                                     |           |                 |
```
Comment (zacharysarah) [inline Documentation/observability/metrics.rst:381]: Capitalize consistently. For fragments, use periods (or not) consistently
```suggestion
|``family``   |Marks which protocol family (IPv4, IPV6) the metric is related to                                   |           |                 |
+-------------+-----------------------------------------------------------------------------------------------------+-----------+-----------------+
|``name``     |Marks a unique identifier for this metric. The name should be defined once for a given type of error.|           |                 |
+-------------+-----------------------------------------------------------------------------------------------------+-----------+-----------------+
|``protocol`` |Marks the L4 protocol (TCP, ANY) for the metric                                                     |           |                 |
```

File: Documentation/observability/metrics.rst [hunk 6]
```diff
@@ -219,255 +219,474 @@ disable the following two metrics as they generate too much data:
...
++-------------+-----------------------------------------------------------------------------------------------------+-----------+-----------------+
+|``signal``   |Marks the signal name                                                                                |           |                 |
++-------------+-----------------------------------------------------------------------------------------------------+-----------+-----------------+
+|``status``   |The label from completed task                                                                        |           |                 |
```
Comment (zacharysarah) [inline Documentation/observability/metrics.rst:385]: Too fragmented
```suggestion
|``status``   |The label from the completed task                                                                        |           |                 |
```

File: Documentation/observability/metrics.rst [hunk 7]
```diff
@@ -219,255 +219,474 @@ disable the following two metrics as they generate too much data:
...
+|``error``  |Indicates the type of error (string)                                                                                                                                                     |           |                 |
++-----------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------+-----------------+
+|``kind``   |The kind of a label                                                                                                                                                                      |           |                 |
++-----------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------+-----------------+
+|``outcome``|Indicates whether the outcome of the operation was successful or not                                                                                                                     |``fail``   |                 |
```
Comment (zacharysarah) [inline Documentation/observability/metrics.rst:493]: Use active voice
```suggestion
|``outcome``|Indicates whether the operation succeeded or not                                                                                                                     |``fail``   |                 |
```

File: Documentation/observability/metrics.rst [hunk 8]
```diff
@@ -219,255 +219,474 @@ disable the following two metrics as they generate too much data:
...
++--------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------+---------------------------------------------------+
+|``enforcement``     |The label used to see the enforcement status                                                                                                                                             |           |                                                   |
++--------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------+---------------------------------------------------+
+|``family``          |Marks which protocol family (IPv4, IPV6) the metric is related to.                                                                                                                       |           |                                                   |
```
Comment (zacharysarah) [inline Documentation/observability/metrics.rst:582]: Use periods (or not) consistently
```suggestion
|``family``          |Marks which protocol family (IPv4, IPV6) the metric is related to                                                                                                                       |           |                                                   |
```

File: Documentation/observability/metrics.rst [hunk 9]
```diff
@@ -219,255 +219,474 @@ disable the following two metrics as they generate too much data:
...
++--------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------+---------------------------------------------------+
+|``protocol``        |Marks the L4 protocol (TCP, ANY) for the metric.                                                                                                                                         |           |                                                   |
++--------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------+---------------------------------------------------+
+|``protocol_l7``     |The label used when working with layer 7 protocols.                                                                                                                                      |           |                                                   |
```
Comment (zacharysarah) [inline Documentation/observability/metrics.rst:590]: Use periods (or not) consistently
```suggestion
|``protocol``        |Marks the L4 protocol (TCP, ANY) for the metric                                                                                                                                         |           |                                                   |
+--------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------+---------------------------------------------------+
|``protocol_l7``     |The label used when working with layer 7 protocols                                                                                                                                      |           |                                                   |
```

File: Documentation/observability/metrics.rst [hunk 10]
```diff
@@ -219,255 +219,474 @@ disable the following two metrics as they generate too much data:
...
+|``proxy_redirects``             |``protocol_l7``|Enabled |           |
++--------------------------------+---------------+--------+-----------+
+
+Policy L7 (HTTP/Kafka) Labels
+"""""""""""""""""""""""""""""
+
++---------------+---------------------------------------------------+-----------+-----------------+
+|Name           |Description                                        |Known Value|Value Description|
++===============+===================================================+===========+=================+
+|``protocol_l7``|The label used when working with layer 7 protocols.|           |                 |
```
Comment (zacharysarah) [inline Documentation/observability/metrics.rst:674]: Use periods (or not) consistently
```suggestion
|``protocol_l7``|The label used when working with layer 7 protocols|           |                 |
```

File: pkg/hive/cell/metric.go [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+	return n
+}
+
+func (m *metric[S]) Apply(container container) error {
+	// Provide the supplied constructor, so its return type is directly accessible by cells
+	container.Provide(m.ctor, dig.Export(true))
+
+	// Provide the metrics provider, which will take the return value of the constructor and turn it into a
+	// slice of metrics to be consumed by anyone interested in handling them.
+	container.Provide(m.provideMetrics, dig.Export(true))
```
Comment (zacharysarah) [inline pkg/hive/cell/metric.go:123]: Nittiest of nits, but use periods consistently, even in code comments
```suggestion
	// Provide the supplied constructor, so its return type is directly accessible by cells
	container.Provide(m.ctor, dig.Export(true)).

	// Provide the metrics provider, which will take the return value of the constructor and turn it into a
	// slice of metrics to be consumed by anyone interested in handling them.
	container.Provide(m.provideMetrics, dig.Export(true))
```

File: pkg/metrics/metric/metric.go [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+package metric
+
+import (
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// WithMetadata is the interface implemented by any metric defined in this package. These typically embed existing
+// prometheus metric types and add additional metadata to the definitions which can be used to generate proper metric
```
Comment (zacharysarah) [inline pkg/metrics/metric/metric.go:13]: Capitalize product names unless they're specifically lower-cased; for example, Prometheus vs gRPC.
```suggestion
// Prometheus metric types and add additional metadata to the definitions which can be used to generate proper metric
```

File: pkg/metrics/metric/metric.go [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+	Namespace string
+	Subsystem Subsystem
+	Name      string
+
+	// Help provides information about this metric.
+	//
+	// Metrics with the same fully-qualified name must have the same Help
+	// string.
+	//
+	// This string is included with the data in the prometheus endpoint and should be brief.
```
Comment (zacharysarah) [inline pkg/metrics/metric/metric.go:75]: Capitalize unless this is the name of a specific endpoint
```suggestion
	// This string is included with the data in the Prometheus endpoint and should be brief.
```

File: pkg/metrics/metric/metric.go [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+		parts = append(parts, o.Subsystem.Name)
+	}
+	parts = append(parts, o.Name)
+
+	return strings.Join(parts, "_")
+}
+
+type LabelDescriptions []LabelDescription
+
+// LabelDescription, unlike traditional prometheus metric declarations contains not only the name of a label but also
```
Comment (zacharysarah) [inline pkg/metrics/metric/metric.go:112]: Capitalize
```suggestion
// LabelDescription, unlike traditional Prometheus metric declarations contains not only the name of a label but also
```

File: tools/metricdoctool/metricdoctool.go [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+		if subsys.Name == "" {
+			if subsys.DocName == "" {
+				key = "misc"
+				subsys = metric.Subsystem{
+					Name:    "",
+					DocName: "Misc",
+				}
+			} else {
+				// If we do have a doc-name, then the place it into its own category in the docs, but the metric
+				// name will not be prefixed with anything.
```
Comment (zacharysarah) [inline tools/metricdoctool/metricdoctool.go:74]: Edits for clarity
```suggestion
				// If we do have a doc-name, then place it into its own category in the docs, but the metric
				// name will not be prefixed with anything.
```

File: tools/metricdoctool/metricdoctool.go [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
+	// Merge all label definitions per subsystem
+	for _, mps := range metricPerSubsystem {
+		for _, m := range mps.Metrics {
+			lbls := m.Labels()
+			for _, lbl := range lbls {
+				desc := mps.Labels[lbl.Name]
+				desc.Name = lbl.Name
+
+				// If there are multiple descriptions for a label, pick the longest. If two are the same size, pick the
+				// alphabetically first. This should ensure a consistent outcome
```
Comment (zacharysarah) [inline tools/metricdoctool/metricdoctool.go:96]: Edits for clarity, consistent periods, avoiding soft conditionals
```suggestion
				// If there are multiple descriptions for a label, pick the longest. If two are the same size, pick the
				// first in alphabetical order. This ensures a consistent outcome.
```

**Review by zacharysarah — CHANGES_REQUESTED** [review]
@dylandreimerink 👋🏻 Some edits for clarity and style, otherwise LGTM from a docs perspective.

File: pkg/metrics/metric/histogram.go [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+			histVec.Reset()
+		}
+	}
+
+	cv.metric.SetEnabled(e)
+}
+
+// HistogramOpts are a modified and expanded version of the prometheus.HistogramOpts.
+// https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#HistogramOpts
+type HistogramOpts struct {
```
Comment (bimmlerd) [inline pkg/metrics/metric/histogram.go:186]: How come this doesn't embed prometheus.HistogramOpts but duplicates it instead? Applies to a few of these similar structs.

File: pkg/hive/example/metrics.go [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+)
+
+var exampleMetricsCell = cell.Metric(newExampleMetrics)
+
+type exampleMetrics struct {
+	ExampleCounter    metric.Counter
+	ExampleCounterVec metric.Vec[metric.Counter]
+}
+
+func newExampleMetrics() exampleMetrics {
```
Comment (bimmlerd) [inline pkg/hive/example/metrics.go:18]: I think this deserves some comments explaining what's going on

File: pkg/metrics/registry.go [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+)
+
+var defaultRegistryConfig = RegistryConfig{
+	PrometheusServeAddr: ":9962",
+}
+
+type RegistryConfig struct {
+	// PrometheusServeAddr IP:Port on which to serve prometheus metrics (pass ":Port" to bind on all interfaces, "" is off)
+	PrometheusServeAddr string
+	//
```
Comment (bimmlerd) [inline pkg/metrics/registry.go:30]: nit: either document or remove comment

File: pkg/metrics/middleware_test.go [hunk 20]
```diff
@@ -35,17 +35,20 @@ func (s *MetricsSuite) TestAPIEventsTSHelperMiddleware(c *C) {
...
 			Next:      http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { w.WriteHeader(test.statusCode) }),
 			TSGauge:   gauge,
 			Histogram: hist,
 		}
 		middleware.ServeHTTP(httptest.NewRecorder(), req)
-		v := testutil.ToFloat64(gauge.WithLabelValues(LabelEventSourceAPI, "/v1/endpoint", http.MethodGet))
+		v := gauge.WithLabelValues(LabelEventSourceAPI.Name, "/v1/endpoint", http.MethodGet).Get()
 		c.Assert(v >= float64(time.Now().Unix()), Equals, test.expectEvent)
-		c.Assert(testutil.CollectAndCount(hist, "test_api_hist") == 1, Equals, test.expectEvent)
+		// TODO FIX
```
Comment (bimmlerd) [inline pkg/metrics/middleware_test.go:50]: todo: fixing test

File: pkg/hive/cell/metric.go [hunk 21]
```diff
@@ -8,9 +8,10 @@ import (
 	"reflect"
 	"runtime"
 
+	"go.uber.org/dig"
```
Comment (bimmlerd) [inline pkg/hive/cell/metric.go:11]: is this the better place for this dep? I thought we only distringuish std&local

**Review by bimmlerd — APPROVED** [review]
It's pretty hard to review this, even commit by commit, as the changes are so large.

Contributing part looks ok to me, though the last commit should probably be squashed

File: pkg/metrics/metric/counter.go [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+
+	Get() float64
+}
+
+type counter struct {
+	prometheus.Counter
+	metric
+}
+
+func (c *counter) Describe(descChan chan<- *prometheus.Desc) {
```
Comment (tommyp1ckles) [inline pkg/metrics/metric/counter.go:33]: Could we have the underlying Counter type be composed with counter, that way we don't have to re implement these functions (and you can still override for Collect()).

**Comment by tklauser** [comment]
There are some open review comments and this branch has picked up quite a few merge conflicts and thus needs a rebase.

I'm going to convert the PR to draft status so it doesn't show up the assigned reviewers' review queue. Please feel free to move out of draft once the above is addressed.

**Comment by dylandreimerink** [comment]
Closing in favor of #25651

**Pull request closed without merging**